### ISSUE TITLE
X11 input broken

### ISSUE THREAD

**Issue by zedeus** [etternagame/etterna#655]
After #643 input is partly broken. Inputs mostly work fine in the menus, like up/down etc, but some double key inputs don't (eg. double up/down to change difficulty), while others do (opening profile options with double enter). When playing a chart inputs are visibly registered but it always misses. Removing the `timer` argument from this line fixes it, but that essentially undos the PR.
https://github.com/etternagame/etterna/blob/4cf772eb9753c0adaf0c2086c85aacb618a867d3/src/arch/InputHandler/InputHandler_X11.cpp#L434

I guess it might be worth noting that this if-statement is always false with the timer argument.
https://github.com/etternagame/etterna/blob/4cf772eb9753c0adaf0c2086c85aacb618a867d3/src/arch/InputHandler/InputHandler.cpp#L18-L25

**Comment by poco0317** [comment]
The if statement in question is meant to kind of produce a timestamp half way between the current update and the previous if no timestamp is given (hence the whole "timestamps bound to framerate" thing). If a timestamp is given, then that doesn't need to happen. That kind of breaks the behavior of `m_iInputsSinceLastUpdate` I guess. Maybe.

Missing in gameplay despite visibly pressing the arrows is very strange to me. The only thing I can think of that would cause that is `INPUTFILTER->ButtonPressed( ... )` coming back true yet having no button press. (to test this, you may not be able to press any single taps but you would still be able to finish a Hold or hit a Mine)

I'm interested in debugging this by knowing what the output of `event.xkey.time` is. It could somehow be possible that the way the timestamp is constructed here is wrong. https://github.com/etternagame/etterna/blob/4cf772eb9753c0adaf0c2086c85aacb618a867d3/src/arch/InputHandler/InputHandler_X11.cpp#L431-L432

In Windows' case of DirectInput, these timestamps are given and that block of code is never entered.

I can't test anything myself outside of Windows.

**Comment by zedeus** [comment]
Mines work but Holds don't.

`event.xkey.time` is in milliseconds like the comment says, here are some examples:
```
2713654334
2713654417
2713654676
2713654755
2713655250
2713655314
2713655333
2713655385
```

I'm not 100% sure and it's probably not relevant, but I think these timestamps are relative to when the X11 server was started.

**Comment by poco0317** [comment]
Double pressing up/down to change difficulty in the menus is bound to CodeDetector. If you have any gameplay buttons bound to those, it wouldn't work (index setup for example). If you happen to have the keys bound that way that would explain that.

I'm not sure how to continue here since none of us have Linux to test on (some of us might but surely I'm the only one who cares enough to even try to fix it). The best I can do is revert the change and wait for a proper solution to be found.

**Comment by zedeus** [comment]
Just using default keybindings + `dfjk`. Reverting is fine with me, maybe @192314 can help?

**Comment by poco0317** [comment]
Linux gang awaken @jameskr97 @kangalioo @Miz0re @Void345 @PPProviden

**Comment by 192314** [comment]
I don't have this issue when building the most recent develop branch (https://github.com/etternagame/etterna/commit/4cf772eb9753c0adaf0c2086c85aacb618a867d3); input works fine for me as far as I can tell.

I'm running Arch, with X Server version 1.18.4.
Is there any other information I could provide to try to debug this issue?

**Comment by Francesco149** [comment]
can confirm this. the weird issue is that initially it worked fine then out of nowhere it stopped taking up/down inputs. did not rebuild or anything.

**Comment by Francesco149** [comment]
actually this could be some kind of timer overflow as my machine is now at over 50 days uptime which is just above the 4 billion millisecond mark. if X11 time is 32-bit unsigned this would be an overflow.

**Comment by Francesco149** [comment]
yes i just made it print the X11 timestamp and it's exactly the value you would get if you truncate my uptime to a 32-bit integer:

![](https://i.imgur.com/Z02TboR.png)

I guess we gotta find a way to not rely on the timestamp to be absolute

**Comment by Francesco149** [comment]
simply using `std::chrono::steady_clock::now` instead of the X11 timestamp fixes the issue, though it's not the most reliable solution because you don't know if the platform actually has a high enough resolution timer

**Comment by zedeus** [comment]
Timer overflow sounds correct, I noticed yesterday etterna's input suddenly broke out of nowhere, and I leave my computer running for weeks at a time as well. After a reboot it worked again

**Comment by nico-abram** [comment]
@zedeus / @Francesco149 Do you know if this is still an issue?

**Comment by dakrk** [comment]
Can confirm this is still an issue. Just noticed this recently and only now did I think it could be uptime related since I have my computer running for months at a time.
